### PR TITLE
fix(tracing): trace frame

### LINKED ISSUE #10845: [BUG] Snapshotter does not replaces frame src attr
**Context:**
- Playwright Version: 1.17.1(also reproduced on `main` branch)
- Operating System: Mac
- Node.js version: v14.15.3
- Browser: All

**Code Snippet**

```javascript
const {chromium, webkit, firefox} = require('playwright');

(async () => {
  const browser = await chromium.launch();
  const ctx = await browser.newContext();
  await ctx.tracing.start({ screenshots: true, snapshots: true });
  const page = await ctx.newPage();
  await page.goto(`http://abehiroshi.la.coocan.jp/`);
  await ctx.tracing.stop({ path: path.join(process.cwd(), "trace.zip") });
  await browser.close();
})();
```

**Describe the bug**
frame `src` attribute is not replaced by snapshotted version. so snapshotted frame content is not shown on trace viewer.

### PR DESCRIPTION
**Pull request by musou1500**
Fixes: #10845

### CODE REVIEW

File: packages/playwright-core/src/server/trace/recorder/snapshotterInjected.ts [hunk 1]
```diff
@@ -419,6 +419,8 @@ export function frameSnapshotStreamer(snapshotStreamer: string) {
               continue;
             if (nodeName === 'IFRAME' && name === 'src')
               continue;
+            if (nodeName === 'FRAME' && name === 'src')
```
Comment (dgozman) [inline packages/playwright-core/src/server/trace/recorder/snapshotterInjected.ts:422]: Let's combine this with previous check like `(nodeName === 'IFRAME' || nodeName === 'FRAME')`. If you rebase on top of the fresh "main" branch, that will make more sense.

File: packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts [hunk 2]
```diff
@@ -87,10 +87,10 @@ export class SnapshotRenderer {
...
     const hideAllStyle = '<style>*,*::before,*::after { visibility: hidden }</style>';
     const prefix = snapshot.doctype ? `<!DOCTYPE ${snapshot.doctype}>` + hideAllStyle : hideAllStyle;
-    html = prefix + html + `
-      <style>*[__playwright_target__="${this.snapshotName}"] { background-color: #6fa8dc7f; }</style>
-      <script>${snapshotScript()}</script>
-    `;
+    html = prefix + [
+      `<style>*[__playwright_target__="${this.snapshotName}"] { background-color: #6fa8dc7f; }</style>`,
+      `<script>${snapshotScript()}</script>`
+    ].join('') + html;
```
Comment (dgozman) [inline packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts:93]: Why this change?

File: packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts [hunk 3]
```diff
@@ -252,7 +251,11 @@ function snapshotScript() {
       }
       document.styleSheets[0].disabled = true;
     };
+
+    const onDOMContentLoaded = () => visit(document);
```
Comment (dgozman) [inline packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts:255]: Why this change?

File: tests/snapshotter.spec.ts [hunk 4]
```diff
@@ -211,14 +231,32 @@ it.describe('snapshots', () => {
...
 });
 
+
+async function waitUntilPass(fn: (counter: number) => Promise<unknown>, sleep = 250) {
+  let counter = 0;
+  for (; ; ++counter) {
+    let err: unknown;
+    try {
+      await fn(counter);
+    } catch (e) {
```
Comment (dgozman) [inline tests/snapshotter.spec.ts:241]: I don't really like that we rely on the exception - who knows what else could throw in this function? Can we do `captureSnapshotWithFrames` helper that would only do the frames trickery?

**Review by dgozman — COMMENTED** [review]
Very nice! A few comments below.

File: tests/snapshotter.spec.ts (hunk 4 above)
Comment (musou1500) [reply tests/snapshotter.spec.ts:241]: Is `captureSnapshotWithFrames` like following?

```typescript
async function captureSnapshotWithFrames(snapshotter: InMemorySnapshotter, page: Page, toImpl:(rpcObject: any) => any) {
  for (let counter = 0; ; ++counter) {
    const snapshot = await snapshotter.captureSnapshot(toImpl(page), 'snapshot' + counter);
    const text = distillSnapshot(snapshot).replace(/frame@[^"]+["]/, '<id>"');
    if (text === '<IFRAME __playwright_src__=\"/snapshot/<id>\"></IFRAME>')
      break;
    await page.waitForTimeout(250);
  }
}
```

**Review by musou1500 — COMMENTED** [review]
Thank you for reviewing. I wrote some comments.

File: tests/snapshotter.spec.ts (hunk 4 above)
Comment (dgozman) [reply tests/snapshotter.spec.ts:241]: Yeah, something like that. Or you can just duplicate this code in both tests, that's totally fine and probably easier than doing some helpers.

File: packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts (hunk 2 above)
Comment (dgozman) [reply packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts:93]: I see, thank you for the explanation. You can now merge hideAllStyle with this `<style>` tag - would be easier to follow.

**Review by dgozman — COMMENTED** [review]
Looks good, let's figure our the tests and simplify a bit.

**Comment by musou1500** [comment]
thank you. I fixed.
considering `distilSnapshot` implementation, I added new `hideAllStyle` style tag to array instead of add styles into existing style tag.

**Comment by dgozman** [comment]
@musou1500 This looks great, but many bot are unhappy. I suspect that's not your changes. Could you please rebase on top of current `main` branch and push, so I'll rerun the tests.

**Comment by musou1500** [comment]
@dgozman I rebased.

**Comment by dgozman** [comment]
Everything looks good, merging.

**Pull request merged**